### PR TITLE
3.1.0 from git source

### PR DESCRIPTION
**Pull request by tomjakubowski**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

This switches to a git-source build mimicking as closely as possible the build done in perspective's CI.

This includes a number of patches which I will work on upstreaming to perspective:
- remove various postinstall hooks (maybe not desirable upstream)
- use header-only boost
- support non-abi3 Python build (until https://github.com/conda/ceps/pull/86 is accepted)
- explicit selection of python interpreter for Maturin
- use conda-provided protoc (protobuf-src crate doesn't build in cross-compiling conda environment)
- patch arrow dependency to solve libtool issue on cross-compiling mac builders
- patch perspectve-server's cmake invocation to use CMAKE_ARGS from environment

### CODE REVIEW

**Comment by tomjakubowski** [comment]
@conda-forge-admin, please rerender

**Comment by tomjakubowski** [comment]
Re-rendering added a whole new dimension now in the build matrix for Node.JS version: 20 or 22. I guess this is because nodejs was added as a build dependency.

The build artifacts should be indifferent to the installed Node.JS version - Node is only used when building the package. I'd like to understand better if we should remove this extra Node.JS dimension or keep it. Node.JS version shouldn't matter for users of the package, but it is still perhaps useful to verify in CI that the package successfully builds with newer Node.JS versions. Digesting all of this now https://docs.conda.io/projects/conda-build/en/stable/resources/variants.html

As for the jupyter extensions, the plan is to fix upstream issues in the sdist, and then update the recipe here to pull from that. That may come in 3.1.1.

**Comment by tomjakubowski** [comment]
@conda-forge-admin, please rerender

**Comment by tomjakubowski** [comment]
`build/force_ignore_keys: [nodejs]` removed nodejs from the build hash. I tried another way, which was [better-attested], by specifying `nodejs: [20]` and `ignore_version: [nodejs]` in conda_build_config.yaml, but local rerenders with that change failed to update the .ci_support files. I'm probably misunderstanding something, will look at the conda-smithy source tomorrow.

[better-attested]: <https://docs.conda.io/projects/conda-build/en/latest/resources/variants.html#special-variant-keys>

**Comment by tomjakubowski** [comment]
Landed on this: removed nodejs as a build dependency -- it comes along with pnpm anyhow -- and that takes it out of the build hash and matrix.

**Comment by tomjakubowski** [comment]
@conda-forge-admin, please rerender

**Comment by tomjakubowski** [comment]
Verified the license file is being packaged:

```
@tomjakubowski ➜ /workspaces/3.0.1 $ tar tvf ../pkg-perspective-python-3.1.0-py312h98163a5_0.tar.zst | grep -i license
-rw-rw-rw- 0/0           10541 2024-10-09 17:26 info/licenses/LICENSE.md
```

Also discovered while doing this that the 3.0.1 packages contain _only_ the license file (note the absence of a pipe to grep).  Makes sense, they're only about a dozen kB: https://anaconda.org/conda-forge/perspective-python/files

```
@tomjakubowski ➜ /workspaces/3.0.1 $ tar tvf pkg-perspective-python-3.0.1-py311h7df6f3f_1.tar.zst 
-rw-rw-r-- 0/0           10541 2024-08-26 02:52 info/licenses/perspective-cpp/LICENSE.md
```

**Comment by tomjakubowski** [comment]
Silly me - the bulk of the package was in the `perspective` package. This PR combines them into one https://anaconda.org/conda-forge/perspective/files

File: recipe/meta.yaml [hunk 1]
```diff
@@ -1,88 +1,76 @@
-{% set name = "perspective" %}
-{% set version = "3.0.1" %}
+{% set name = "perspective-python" %}
```
Comment (timkpaine) [inline recipe/meta.yaml:1]: Undo

**Review by timkpaine — CHANGES_REQUESTED** [review]

File: recipe/meta.yaml (hunk 1 above)
Comment (tomjakubowski) [reply recipe/meta.yaml]: OK. I figured it would be best to match the name in pypi but happy to change

**Comment by tomjakubowski** [comment]
@conda-forge-admin, please rerender

**Comment by tomjakubowski** [comment]
Locally, mac builds with `ARROW_DEPENDENCY_SOURCE=BUNDLED` are building with references to jemalloc symbols in libarrow.a, despite Arrow's build reporting `ARROW_JEMALLOC=OFF`. This causes link errors from the .so on importing perspective, which trips the import test. I can also replicate this misconfigured build in perspective outside of conda, ∑hen I apply the patch to use `ARROW_DEPENDENCY_SOURCE=BUNDLED`.

I don't know what this is yet, but it does appear that the import test are succeeding in CI, so it is possible it is something particular about my local setup. May try someone else as a guinea pig

File: recipe/patches/fix_arrow_libtool.patch [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+--- a/cmake/arrow.txt.in
++++ b/cmake/arrow.txt.in
+@@ -9,8 +9,8 @@ project(arrow-download NONE)
+
+ include(ExternalProject)
+ ExternalProject_Add(apachearrow
+-  GIT_REPOSITORY    https://github.com/apache/arrow.git
+-  GIT_TAG           apache-arrow-17.0.0
++  GIT_REPOSITORY    https://github.com/ProspectiveCo/arrow.git
++  GIT_TAG           tom-libtool-fix2
```
Comment (timkpaine) [inline recipe/patches/fix_arrow_libtool.patch:12]: No

**Review by timkpaine — CHANGES_REQUESTED** [review]

File: recipe/patches/fix_arrow_libtool.patch (hunk 2 above)
Comment (tomjakubowski) [reply recipe/patches/fix_arrow_libtool.patch:12]: That tag includes [a patch](https://github.com/apache/arrow/pull/44385/commits/f2b12a311d1a14ab5d9f67428201e1389748a9c1) to work around an [obscure issue](https://github.com/apache/arrow/issues/44384) in Arrow when building it with bundled dependencies when /usr/bin/libtool is broken (as it seems to be on the conda-forge cross-compiling Mac builders). 

The bespoke tag on the Prospective fork was an expedient way to test that change turned builds green. Aside from removing the Mac builders and waiting for an arrow release which includes https://github.com/apache/arrow/pull/44385/, the alternative is to patch arrow.txt.in to add a `PATCH_COMMAND` which includes the arrow patch. I'll do that.

Comment (tomjakubowski) [reply recipe/patches/fix_arrow_libtool.patch]: Patch isn't sufficient, need to make perspective's build read `CMAKE_ARGS` from the environment too. In the python build, cmake is invoked from perspective-server's build script using the `cmake` crate.  This is where it's rocky: the cmake crate doesn't support passing arbitrary arguments to cmake. As of 2017 they said they'd accept a PR to add the feature. [open issue](https://github.com/rust-lang/cmake-rs/issues/27)

Two ways occur to me solve this:
- parse CMAKE_ARGS in the Rust build script from the environment and turn each parsed arg into a method call on the cmake::Config builder object, e.g. `cmake.define(var, var)` which would define
- parse CMAKE_ARGS from the environment in CMakeLists.txt and turn them into`set(${var} ${value})` and other cmake calls

could also side-step this and just patch the the cmake invocation in FindInstallDependency.cmake to pass on `-DCMAKE_LIBTOOL=$ENV{LIBTOOL}` but I would rather make perspective's build respect all CMAKE_ARGS if it isn't too much work

**Pull request closed without merging**